plugin/storage/clickhouse/factory.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2023 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package clickhouse
+
+import (
+	"context"
+	"database/sql"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/pkg/metrics"
+	chSpanStore "github.com/jaegertracing/jaeger/plugin/storage/clickhouse/spanstore"
+	"github.com/jaegertracing/jaeger/storage/dependencystore"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+type Factory struct {
+	client         *sql.DB
+	spansTableName string
+	logger         *zap.Logger
+}
+
+func NewFactory(ctx context.Context, cfg Config, logger *zap.Logger) *Factory {
+	f := &Factory{
+		logger:         logger,
+		spansTableName: cfg.SpansTableName,
+	}
+
+	client, err := cfg.NewClient(ctx)
+	if err != nil {
+		f.logger.Error("failed to create ClickHouse client", zap.Error(err))
+	} else {
+		f.client = client
+	}
+
+	return f
+
+	//TODO: Move some steps to Initialize()
+}
+
+func (f *Factory) CreateSpansTable(ctx context.Context) {
+	if err := chSpanStore.CreateSpansTable(ctx, f.client, f.spansTableName); err != nil {
+		f.logger.Error("failed to create spans table", zap.Error(err))
+	}
+}
+
+func (f *Factory) ExportSpans(ctx context.Context, td ptrace.Traces) error {
```

Comment:
why is ExportSpans defined on the factory instead of span store?

Reply:
The logic is defined in `plugin/storage/spanstore/clickhouse/exporter.go` ([link](https://github.com/jaegertracing/jaeger/pull/4941/files#diff-e85bace6bfd6e03dd602c1d1bd2ec27367e539b554e541d6f30786a63d652f2eR43)). But the exporter is depending on the real type of Factory to decide whether to call `SpanWriter.WriteSpan()` or `ExportSpans()` ([link](https://github.com/jaegertracing/jaeger/pull/4941/files#diff-d11f17877ed77e35143c698010924340c6cf95489656e3ccb0b3c2654c9b9853R42-R52). So I think making `ExportSpans()` a func of clickhouse Factory may be a good idea.

Of course, a better idea is to define a new API for `ExportSpans()`, but maybe we can do that later? Or you have a better idea?

Reply:
I have a better idea, but it is not well formed yet, I need to write it down and maybe put together the new interfaces. Will try to do this later this weekend. Are you blocked on it?

Reply:
No I don't think this blocks the progress. Can continue.